SweepPoints.get_meas_obj_sweep_points_map

stephlazar opened the pull request:
SweepPoints.get_meas_obj_sweep_points_map: rewrites this function to extract all the sweep parameters in each dimension for each meas_obj.

Before the PR, this method would only take the first sweep parameter in each dimension.

Commit:
SweepPoints.get_meas_obj_sweep_points_map: rewrites this function to extract multiple parameters per meas_obj

Commit:
SweepPoints.get_meas_obj_sweep_points_map: adds comments and improved docstring

chellings commented on pycqed/measurement/sweep_points.py:
```diff
@@ -270,44 +270,59 @@ def get_meas_obj_sweep_points_map(self, measurement_objects):
+
+        :param measurement_objects: list of strings to be used as keys in the
+            returned dictionary. These are the measured object names.
+            Can also be list of QuDev_transmon instanced in which case
```
The comment can be written in a more general way: in the future, we might have other mobjs that have a name attribute.

stephlazar replied:
I changed this now

chellings commented on pycqed/measurement/sweep_points.py:
```diff
@@ -270,44 +270,59 @@ def get_meas_obj_sweep_points_map(self, measurement_objects):
+
         for i, mobj in enumerate(measurement_objects):
             if hasattr(mobj, 'name'):
+                # list of QuDev_transmon instances was provided
```
Same as above: the comment can be written in a more general way: in the future, we might have other mobjs that have a name attribute.

stephlazar replied:
Changed now

chellings commented on pycqed/measurement/sweep_points.py:
```diff
@@ -270,44 +270,59 @@ def get_meas_obj_sweep_points_map(self, measurement_objects):
+                all_pars = []
+                for mobjn in measurement_objects:
+                    # Find all sweep param names that contain the mobj name
+                    pars = [k for k in list(d) if mobjn in k]
```
It seems to me that this could introduce bugs like adding qb12 sweep points for qb1 (or even weirder bugs if someone comes up with completely different mobj names). Shouldn't it be `if k.startswith(mobjn + '_')`?

stephlazar replied:
Ooops! I had already introduced this bug before somewhere else--good catch! :p 
But I replaced it with `k.startswith(mobj)` without the underscore because the user might now always use QE to define sweep parameter name and might have something like "qb12amplitude"

chellings replied:
Without the underscore, it would still add qb12 sweep points for qb1.

stephlazar replied:
Indeed ...
I wonder if you approve the solution I came up with now (after spending 2 hours ... ). In the end I don't know how to make it completely general, so now the only compromise is that it cannot find sweep parameter names that contain more than one mobjn (for example 'qb1qb9_cliffords'), which was true even in the solution using `k.startswith(mobjn + '_')`.

So now the function can handle correctly both of the cases below:

```
sweep_params_1d = ['this_qb15_amplitude1', 'this_qb11_freq1', 'this_qb1_amplitude1', 'this_qb3_freq1']
sweep_params_2d = ['this_qb15_amplitude2', 'this_qb11_freq2', 'this_qb1_amplitude2', 'this_qb3_freq2']
mobjns = ['qb3', 'qb1', 'qb11', 'qb15']
swpts.get_meas_obj_sweep_points_map(mobjns) gives
{'qb3': ['this_qb3_freq1', 'this_qb3_freq2'],
 'qb1': ['this_qb1_amplitude1', 'this_qb1_amplitude2'],
 'qb11': ['this_qb11_freq1', 'this_qb11_freq2'],
 'qb15': ['this_qb15_amplitude1', 'this_qb15_amplitude2']}
```


```
sweep_params_1d = ['this_twpaA_amplitude1', 'this_twpa_freq1', 'thistwpaB_freq1']
sweep_params_2d = ['this_twpaA_amplitude2', 'this_twpa_freq2', 'thistwpaB_freq2']
mobjns = ['twpa', 'twpaA', 'twpaB']
swpts.get_meas_obj_sweep_points_map(mobjns) gives
{'twpa': ['this_twpa_freq1', 'this_twpa_freq2'],
 'twpaA': ['this_twpaA_amplitude1', 'this_twpaA_amplitude2'],
 'twpaB': ['thistwpaB_freq1', 'thistwpaB_freq2']}
```

chellings reviewed (commented):
A question since I am not fully sure about the implications of these changes. Is there analysis code on qudev_master (e.g., in tda) that expects the old behavior of this function? If yes, should we include the respective changes in analysis into this PR? We could also decide to merge the new version of this function as part of a larger PR (probably timedomain classes?) if it makes more sense to merge it together with other changes.

Please also note the further comments.

Commit:
SweepPoints.get_meas_obj_sweep_points_map: fixes bug related to qb name string recognition and improves docstring + comments to not say QuDev_transmon

- the bug was related to failing to distinguish 'qb1' and 'qb1x'

Commit:
SweepPoints.get_meas_obj_sweep_points_map: proper fix to distinguish between, for ex 'qb1' and 'qb12' in the sweep parameter names

Now the functions does the following:
- sorts measurement_objects by length
- only concludes that mobjn is in the sweep param name if longer mobjn in measurement_objects are not in the sweep param name

Commit:
Merge branch 'qudev_master' into enh/sweep_points

Commit:
SweepPoints.get_meas_obj_sweep_points_map: changes function to look for mobjn in sweep_param_name.split('_')

stephlazar commented:
I now changed the function to find mobjn in k.split('_'), as discussed on Slack. A warning in the docstring informs the user that this is assumed.

chellings reviewed (approved):
Thank you!
